advanced_source/cpp_autograd.rst

```diff
@@ -0,0 +1,442 @@
+either by putting ``torch::NoGradGuard`` in a code block
+
+.. code-block:: cpp
+
+  std::cout << x.requires_grad() << std::endl;
+  std::cout << x.pow(2).requires_grad() << std::endl;
+  
+  {
+    torch::NoGradGuard no_grad;
+    std::cout << x.pow(2).requires_grad() << std::endl;
+  }
+
+
+Out:
+
+.. code-block:: shell
+
+  true
+  true
+  false
+
+Or by using ``.detach()`` to get a new Tensor with the same content but that does
+not require gradients:
+
+.. code-block:: cpp
+
+  std::cout << x.requires_grad() << std::endl;
+  y = x.detach();
+  std::cout << y.requires_grad() << std::endl;
+  std::cout << x.eq(y).all().item<bool>() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+  true
+  false
+  true
+
+For more information on C++ tensor autograd APIs such as ``grad`` / ``requires_grad`` /
+``is_leaf`` / ``backward`` / ``detach`` / ``detach_`` / ``register_hook`` / ``retain_grad``,
+please see `the corresponding C++ API docs <https://pytorch.org/cppdocs/api/classat_1_1_tensor.html>`_.
+
+Computing higher-order gradients in C++
+---------------------------------------
+
+One of the applications of higher-order gradients is calculating gradient penalty.
+Let's see an example of it using ``torch::autograd::grad``:
+
+.. code-block:: cpp
+
+  #include <torch/torch.h>
+  
+  auto model = torch::nn::Linear(4, 3);
+  
+  auto input = torch::randn({3, 4}).requires_grad_(true);
+  auto output = model(input);
+  
+  // Calculate loss
+  auto target = torch::randn({3, 3});
+  auto loss = torch::nn::MSELoss()(output, target);
+  
+  // Use norm of gradients as penalty
+  auto grad_output = torch::ones_like(output);
+  auto gradient = torch::autograd::grad({output}, {input}, /*grad_outputs=*/{grad_output},
+                  /*retain_graph=*/true, /*create_graph=*/true,
+                  /*allow_unused=*/true)[0];
+  gradient = gradient.view({-1, 1});
+  auto gradient_penalty = torch::pow((gradient.norm(2, /*dim=*/1) - 1), 2).mean();
+  
+  // Add gradient penalty to loss
+  auto combined_loss = loss + gradient_penalty;
+  combined_loss.backward();
+  
+  std::cout << input.grad() << std::endl;
+
+Out:
+
+.. code-block:: shell
+
+  -0.2574  0.0494  0.3060  0.3215
+  -0.1197  0.0248  0.1405  0.1401
+  [ CPUFloatType{2,4} ]
+
+Please see the documentation for ``torch::autograd::backward``
+(`link <https://pytorch.org/cppdocs/api/function_namespacetorch_1_1autograd_1afa9b5d4329085df4b6b3d4b4be48914b.html>`_)
+and ``torch::autograd::grad``
+(`link <https://pytorch.org/cppdocs/api/function_namespacetorch_1_1autograd_1a1e03c42b14b40c306f9eb947ef842d9c.html>`_)
+for more information on how to use them.
+
+Using custom autograd function in C++
+-------------------------------------
+
+(Adapted from `this tutorial <https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd>`_)
+
+Adding operations to ``torch::autograd`` requires implementing a new ``torch::autograd::Function``
```

Comment:
"operations" here might not be the best naming (even though it is in the other tutorial).
Maybe "Adding a new elementary operation to"